### PR TITLE
Remove `ThriftType` enum

### LINKED ISSUE #72: workload::Value and request::thrift::ThriftValue are almost identical
Something I've run into hacking on #64 is that the [workload::Value](https://github.com/twitter/rpc-perf/blob/master/lib/workload/src/lib.rs#L116) enum and [request::thrift::ThriftValue](https://github.com/twitter/rpc-perf/blob/master/lib/request/src/thrift.rs#L34) enum are nearly identical. This results in tree translation [here](https://github.com/twitter/rpc-perf/blob/master/lib/workload/src/lib.rs#L358). By using the same tree structure the copy could be avoided. I'm suggesting making `Parameter` and `Value` the 'tree to rule them all' which can be directly used by things like `thrift` for generating a buffer. IMO this makes sense because all the information harvested from the config parsing should be available in this tree, rendering everything else strictly a subset.

A complication is that the build is split out in such a way that `rpcperf_workload` depends on `rpcperf_request` but the tree is defined in `rpcperf_workload` and it results in a cyclic dependency when trying to expose this tree to both modules. Two possible solutions for this are to make yet another module just for that tree or merge rpcperf_workload and rpcperf_request modules. The workload module is a single file so I think merging them is the easiest thing to do.

On the other hand, maybe this is not a good idea altogether for reasons I've not yet contemplated.

### PR DESCRIPTION
**Pull request by bryce-anderson**
closes #72 

This PR removes the `ThriftType` tree. Doing so required merging the `workload` module with the `request` module. In principle they were very strongly coupled so it is probably for the best.

### CODE REVIEW

**Comment by bryce-anderson** [comment]
This is a work in progress. 
Things to address: 
* [ ] Type `Double` isn't implemented for thrift
* [ ] Some of the types in the `workload` module might be best in the base `request` module
* [ ] `ThriftRequest<'a>` is not all that useful. Maybe it can be axed?

**Comment by bryce-anderson** [comment]
About the panics: I think the handling of missing field id's may be incorrect but sense there is no official spec I'm not certain. It looks like in code generated from IDL, fields that don't have a hard coded id are given one starting and -1 and decremented with each additional unspecified field. Currently the field and type codes are simply omitted which probably results in a corrupt buffer.

If my analysis is correct, I think that would be something for another issue after this is merged.

**Comment by bryce-anderson** [comment]
Right now this PR addresses the goals of #72, but some of the points enumerated above deserve a bit of discussion. 
* I think point 2 is probably unnecessary for now. 
* Point 3 would be good but is not a priority.
* As noted above, I think point 4 is a protocol error waiting to happen and deserves a separate issue/PR.

**Comment by brayniac** [comment]
Point 2 - agree, not necessary right now
Point 3 - the `ThriftRequest` is certainly less useful now that the crates are merged. I'm not worried about it
Point 4 - this is intentional actually.. for instance if you have a `List` of `String`it serializes as the type code for list, the type code for the contained type, and the length, followed by a.len() a b.len() b. Essentially, in the config, parameters without `id` are not treated as fields. That is, no field id or type code are written to the buffer, just the serialized form of the value. Once #64 is addressed, this should go away too? I know the thrift calc example is too basic to demonstrate this, but I'm relying on this functionality internally, and it is working as expected.

File: lib/request/src/thrift.rs [hunk 1]
```diff
@@ -307,6 +310,7 @@ pub fn generic(request: ThriftRequest) -> Vec<u8> {
                 buffer.write_bytes(&[byte]);
                 buffer.write_i32(len);
             }
+            _ => { /* silently ignored. Not good. */ }
```
Comment (brayniac) [inline lib/request/src/thrift.rs]: Should we `panic!()` here? It's probably worse to send malformed data.

**Comment by bryce-anderson** [comment]
Ah, now point 4 makes sense. I didn't look into how that would interact with collection types. I do think that it would be cleaned up once making the structure recursive.

I'm not sure what you want to do about `ThriftRequest` though it seems like its not important either way. Should I squash these commits (and maybe cut `ThriftRequest`) and we can get this merged?

File: lib/request/src/workload.rs [hunk 2]
```diff
@@ -13,26 +13,18 @@
...
-extern crate rand;
-extern crate ratelimit;
-extern crate rpcperf_request as request;
-extern crate shuteye;
-extern crate time;
-
 const ONE_SECOND: u64 = 1_000_000_000;
 pub const BUCKET_SIZE: usize = 10_000;
 
+// Why do I need self?
```
Comment (brayniac) [inline lib/request/src/workload.rs]: ?

**Comment by bryce-anderson** [comment]
Sorry, looks like I jumped the gun. I'll give you time to look through it.

File: lib/request/src/thrift.rs (hunk 1 above)
Comment (bryce-anderson) [inline lib/request/src/thrift.rs]: Agreed. The only type that isn't handled is `None` but a panic still seems appropriate.

File: lib/request/src/workload.rs (hunk 2 above)
Comment (bryce-anderson) [inline lib/request/src/workload.rs]: This is an leftover from some confusion I had with the `extern crate` and module system. It should go.

**Comment by brayniac** [comment]
Alright, those are the only comments I have. I think removing the `ThriftRequest` should be pretty quick. I don't see value in keeping it. It's outside the original scope of this PR, so I'll leave it to your judgement. Making `generic()` take `method`, `sequence`, `Vec<Type>` seems just as good.

Otherwise, yes, address the two comments, squash and I'll do a quick few tests before I merge. (We need more test coverage)

**Comment by bryce-anderson** [comment]
I have removed `ThriftRequest` and squashed. Let me know if you find any problems.

File: lib/request/src/thrift.rs [hunk 3]
```diff
@@ -218,86 +191,105 @@ impl Buffer {
...
-    generic(ThriftRequest {
-        method: "ping",
-        sequence_id: 0,
-        payload: Vec::<ThriftType>::new(),
-    })
+    generic("ping", 0, &Vec::new())
 }
 
-pub fn generic(request: ThriftRequest) -> Vec<u8> {
+pub fn generic(method: &str, sequence_id: i32,payload: &Vec<Parameter>) -> Vec<u8> {
```
Comment (brayniac) [inline lib/request/src/thrift.rs]: nit: include a space following the comma, rustfmt should fix this

Comment (bryce-anderson) [inline lib/request/src/thrift.rs]: Done. :)

**Comment by brayniac** [comment]
Looks ready to merge.

**Pull request merged**